smithy-rules-engine/src/main/java/software/amazon/smithy/rulesengine/traits/ContextIndex.java

```diff
@@ -0,0 +1,103 @@
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.knowledge.KnowledgeIndex;
+import software.amazon.smithy.model.node.Node;
+import software.amazon.smithy.model.shapes.MemberShape;
+import software.amazon.smithy.model.shapes.OperationShape;
+import software.amazon.smithy.model.shapes.ServiceShape;
+import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.model.shapes.StructureShape;
+import software.amazon.smithy.model.shapes.ToShapeId;
+import software.amazon.smithy.utils.SmithyUnstableApi;
+
+/**
+ * Resolves an indexes the context parameters in the model.
+ */
+@SmithyUnstableApi
+public final class ContextIndex implements KnowledgeIndex {
+    private final Map<ShapeId, Map<String, ClientContextParamDefinition>> serviceParams = new HashMap<>();
+    private final Map<ShapeId, Map<String, StaticContextParamDefinition>> operationParams = new HashMap<>();
+    private final Map<ShapeId, Map<MemberShape, ContextParamDefinition>> operationMemberParams = new HashMap<>();
+
+    public ContextIndex(Model model) {
+        for (ServiceShape serviceShape : model.getServiceShapes()) {
+            serviceShape.getTrait(ClientContextParamsTrait.class).ifPresent(trait -> {
+                serviceParams.put(serviceShape.toShapeId(), trait.getParameters());
+            });
+        }
+
+        for (OperationShape operation : model.getOperationShapes()) {
+            operation.getTrait(StaticContextParamsTrait.class).ifPresent(endpointParamsTrait -> {
+                operationParams.put(operation.toShapeId(), endpointParamsTrait.getParameters());
+            });
+            operation.getInput().ifPresent(shapeId -> {
+                StructureShape input = model.expectShape(shapeId, StructureShape.class);
+                Map<MemberShape, ContextParamDefinition> parameterReferenceMap = new LinkedHashMap<>();
+                input.members().forEach(memberShape -> {
+                    memberShape.getTrait(ContextParamTrait.class).ifPresent(contextParamTrait -> {
+                        parameterReferenceMap.put(memberShape, ContextParamDefinition.builder()
+                                .name(contextParamTrait.getName())
+                                .build());
+                    });
+                });
+                operationMemberParams.put(operation.toShapeId(), parameterReferenceMap);
+            });
+        }
+    }
+
+    public static ContextIndex of(Model model) {
+        return model.getKnowledge(ContextIndex.class, ContextIndex::new);
+    }
+
+    /**
+     * Gets the mapping of context parameter names and corresponding {@link ClientContextParamDefinition} to be
+     * generated on the service client's configuration.
+     *
+     * @param service The service shape.
+     * @return The mapping of context parameter names to {@link ClientContextParamDefinition}.
+     */
+    public Map<String, ClientContextParamDefinition> getClientContextParams(ToShapeId service) {
```

Comment:
I think for this index, maybe we save the overhead and memory of precomputing this and just return the trait for a `Shape` rather than a `ToShapeId`. Same comment for getStaticContextParams since there's no actual computation.